### ISSUE TITLE
ENH: Generation of random 2D tables with given marginal totals

### ISSUE THREAD

**Issue by HDembinski** [scipy/scipy#15618]
## Motivation

Sometimes you have pairs of values x, y and you want to know whether y is independent of x or not. People sometimes compute a test statistic of x, y pairs directly, examples are Pearson/Spearman coefficient, Kendall tau or similar.

Other tests use contingency tables (histograms), like the popular [USP test](https://royalsocietypublishing.org/doi/10.1098/rspa.2021.0549). This can be more computationally efficient when the sample is very large (think of applications with Big Data, like in my field, high-energy physics). For these tests, one needs to generate random 2D tables that have the same row and column totals as the original data. The probability distribution for such a table has been worked out in the 80'ies, and Patefield found an elegant non-trivial algorithm to randomly sample from this distribution.

### Describe the solution you'd like.

I searched the net for implementations in C/C++ of Patefield's algorithm, but only found implementations under GPL license and none of them is available to Python. I want to have the code under a less restrictive license and so I translated the published FORTRAN code to C from scratch and [based my implementation on that](https://github.com/resample-project/resample/blob/develop/src/rcont.c). My code is licensed under BSD 3-clause and I would like to contribute it to SciPy.

My implementation already evolved from the original and has cool features that the original implementation lacks. The changes are documented in the code. Sorry for all the `goto`s, it is how the original code was written. In time, I am interested in removing the `goto`s step by step, but for now it is more important to have a baseline code that works.

In addition to Patefield's algorithm, I also want to add a simple shuffling algorithm. The shuffling algorithm requires additional memory, but is faster than Patefield in certain scenarios. Patefield is more time efficient when the table contains few cells and depends only logarithmically on the number of total entries, while the time for the shuffling algorithm does not depend on the number of cells and only depends on the number of entries.

I would like to start a PR; some pointers where this algorithm fits in best, would be helpful.

### Describe alternatives you've considered.

As mentioned above, before I started writing mine, there were only GPL licensed implementations and none available to Python.

**Comment by tupui** [comment]
Hi, thanks for the proposal @HDembinski. I am not sure to understand, could you give some concrete numerical examples? Is this similar to [`owens_t`](https://scipy.github.io/devdocs/reference/generated/scipy.special.owens_t.html)?

**Comment by HDembinski** [comment]
owens_t is something different as you can see from the docs of that function. If you look at the links that listed, it should become more clear. I hope. The paper for the USP test was written for non experts, is open access and very accessible.

**Comment by HDembinski** [comment]
Where would be the best place to put `rcont1` and `rcont1` in scipy?

**Comment by tupui** [comment]
> Where would be the best place to put `rcont1` and `rcont2` in scipy?

I am not sure yet if it should be included or not. @mdhaber what do you think?

**Comment by mdhaber** [comment]
Sounds like a multivariate distribution? Many in of them have only `rvs` implemented.
Would you send an email to the [scipy-dev list](https://mail.python.org/mailman3/lists/scipy-dev.python.org/) about this and invite people to this issue for discussion? Then we can make a decision based on wider feedback.

**Comment by mdhaber** [comment]
@HDembinski I do think this could be in scope if it were added as a multivariate distribution. Does that sound reasonable? What would its name be, and what parameters would it have? (It wouldn't necessarily need any methods except for `rvs`, but in principle, other methods could be added.)

**Comment by HDembinski** [comment]
The algorithm only makes sense for 2D matrices, since you have pairs of values that are independent and not tuples.

**Comment by mdhaber** [comment]
But each of the two dimensions can be of any size (in principle), right? Several of the multivariate distributions produce 2D random variates (e.g. [random_correlation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.random_correlation.html#scipy.stats.random_correlation)).

**Comment by mdhaber** [comment]
@HDembinski are you still interested in contributing this? I think it could be quite useful, and I'd be willing to review it if it were implemented in Python, but I'm afraid I can't help with the C. 

I still think it would fit as a multivariate distribution. The random variable here is a 2D matrix. If we can figure out a name for this distribution, I think it would have an `rvs` method that accepts parameters `r` (for row sums), `c` (column sums), `size` (the number of variates to draw), and `random_state` (the RNG object or seed).

Suppose the name were `patefield`. It might be called like:
```python3
rng = np.random.default_rng(3459034)
size = 2
r = [5, 2]
c = [3, 4]
patefield.rvs(r, c, size=size, random_state=rng)
```
which might produce:
```
array([[[3, 1],
        [2, 1]],

       [[4, 0],
        [1, 2]]])
```

**Comment by HDembinski** [comment]
> @HDembinski are you still interested in contributing this? I think it could be quite useful (e.g. for [gh-7099](https://github.com/scipy/scipy/issues/7099)) and I'd be willing to review it if it were implemented in Python, but I'm afraid I can't help with the C.

Cool, thanks for that. Yes, I am still interested. I also convinced the Boost.Math maintainer that it would be good to add these algorithms to Boost.Math, therefore my plan was to add it first to Boost.Math and then this can be wrapped to scipy?

Unless we are in a hurry, then I can try to add it here and in parallel to Boost.Math. I cannot predict how long it will take to get the PR merged into Boost.Math.

> I still think it would fit as a multivariate distribution. The random variable here is a 2D matrix. If we can figure out a name for this distribution, I think it would have an `rvs` method that accepts parameters `r` (for row sums), `c` (column sums), `size` (the number of variates to draw), and `random_state` (the RNG object or seed).

Yeah, that makes sense.

> Suppose the name were `patefield`. It might be called like:
> 
> ```python
> rng = np.random.default_rng(3459034)
> size = 2
> r = [5, 2]
> c = [3, 4]
> patefield.rvs(r, c, size=size, random_state=rng)
> ```
> 
> which might produce:
> 
> ```
> array([[[3, 1],
>         [2, 1]],
> 
>        [[4, 0],
>         [1, 2]]])
> ```

The interface looks good, but I think we should use a more generic name. There are two algorithms that can generate these tables, which one is faster depends on the total number of entries in the table and the number of cells. Patefield is one of them, the other one is a trivial one that I came up with myself (but since it is trivial, I don't claim any intellectual ownership).

**Comment by mdhaber** [comment]
> I think it makes sense to add it first here...
>There are two algorithms that can generate these tables,

OK. If it's going here first, let's start with the trivial algorithm implemented in pure Python. This would allow us to work out the API and get the new class approved on the mailing list, saving the more complicated algorithmic stuff for a follow-up PR.

Does that sound good?

> I think we should use a more generic name.

Sure, I agree. Let's use "table" instead of "matrix" just to avoid any possibility of confusion with `np.matrix`. I think `conditional_table` is apt and concise, but `table_with_marginals` (or `fixed_marginal_table`?) is a little more obvious... I could go either way. Pick your favorite, and we'll propose it on the mailing list once there's a PR for people to look at. Perfect naming is hard; all we can do is pick one that nobody strongly disagrees with.

**Comment by HDembinski** [comment]
@mdhaber I looked into scipy.stats._multivariate and I don't know how to implement all of the methods for a general distribution object. There is not only `rvs`, but also `mean`, `pmf` (and `logpmf`), `var`, `cov`. I know what `mean` should return, but I have no idea about `pmf`, `var`, and `cov`.

**Comment by rkern** [comment]
There are no required methods for the multivariate distributions because they are all much more diverse than the univariate ones (and are often in spaces that aren't Euclidean vector spaces, like here). Implement what makes sense for your distribution.

As for the `pmf`, if I'm reading the references correctly, this distribution is uniform over all of the contingency tables that have the given marginal sums, right? Then the `pmf` would be not particularly useful to anyone; it's either 0 when the marginals aren't met and the reciprocal of the count of valid contingency tables otherwise. Computing that count might be hard, and there isn't really anything useful one could want to do with that PMF (nothing to optimize), so I wouldn't worry about implementing it.

**Comment by WarrenWeckesser** [comment]
For 2x2 tables, the PMF is given by the hypergeometric distribution.  For tables with shape (m, n), the appropriate generalization is needed.  A contingency table is generated by counting unique pairs of some data set.  Each possible data set with the prescribed quantities of categories in the components of the pairs (i.e. with given marginal sums) is equally likely, but multiple data sets can result in the same contingency table.

**Comment by rkern** [comment]
As far as I can tell, the algorithm here samples uniformly from the set of _contingency tables_ with the given marginal sums, not uniformly from the set of datasets with the given marginal sums.

**Comment by HDembinski** [comment]
@rkern @WarrenWeckesser Thanks for the quick feedback! Ok, I will carry on then and implement what I know.

> this distribution is uniform over all of the contingency tables that have the given marginal sums, right?

I am not sure that one can express it that way. I think Warren is right that 2x2 tables are described by the hypergeometric distribution. Another way to describe the set is as a subset of the multinominal distribution. The multinominal distribution requires that the sum of all entries is fixed, while this distribution requires also row and column sums to be fixed. The row and column sums also fully determine the expected frequency for each cell in the table.

Looking at the hypergeometric distribution, there is probably a closed formula for the covariance of this distribution, but I don't know what it is.

**Comment by rkern** [comment]
Okay, yeah, I was looking at some later papers that referenced Patefield that were in fact sampling uniformly from the set of contingency tables. But Patefield is not doing that. In any case, he references an [earlier paper](https://www.jstor.org/stable/2347219) that does give the PMF explicitly.

**Comment by WarrenWeckesser** [comment]
FYI, and only noted as an alternative to which a more efficient implementation can be compared: In the github repo [numerand](https://github.com/WarrenWeckesser/numerand), I have an implementation of a brute-force method for generating random contingency tables.  To generate a random table, it generates an appropriately sized dataset with values that match the given marginals, shuffles the appropriate components to randomize the dataset, and then counts the result.  It works fine for marginal sums that are not very big, but the algorithm is O(N) in memory and time, where N is the sum of the values in any contingency table, so it is not recommended for tables with large marginal sums.  As I wrote in the docstring of `rand_table`, "There is a lot of literature on the efficient generation of random contingency tables that has been completely ignored here!"  It does handle 2x2 tables as a special case, where it uses the hypergeometric distribution to generate the random tables.

Some scratch-work for a C implementation of the same brute-force method is in https://github.com/WarrenWeckesser/experiments/tree/master/c/randtable

**Comment by mdhaber** [comment]
I asked @HDembinski to start with such a brute-force algorithm in Python so we can get the interface approved (i.e. not rejected) by the mailing list. In a follow-up PR, I suggested that we could speed it using Pythran, or we could replace it with the Patefield algorithm, then speed it up with Pythran. If we go with C/C++, I can't help as much.

**Comment by WarrenWeckesser** [comment]
Yes, that's what I have been calling the "brute force" method.  You can also compare `np.mean(tables, axis=0)` to `stats.contingency.expected_freq(tables[0])`.  With `n_tables` increased to `10000`, I get:
```
In [63]: np.mean(tables, axis=0)
Out[63]: 
array([[ 4.1704,  5.8296],
       [ 8.307 , 11.693 ],
       [12.5226, 17.4774]])

In [64]: stats.contingency.expected_freq(tables[0])
Out[64]: 
array([[ 4.16666667,  5.83333333],
       [ 8.33333333, 11.66666667],
       [12.5       , 17.5       ]])
```
The code in `numerand.random_table` uses a somewhat faster cross-tab function that is specialized to this particular case.  It also (by default) uses the hypergeometric distribution for the 2x2 case, which is *much* faster than the brute force method.

**Comment by HDembinski** [comment]
> I have an implementation of a brute-force method for generating random contingency tables.

@WarrenWeckesser Yes, that's the "shuffling algorithm" I refer to in my OP. I have a C implementation of that algorithm [which I use in https://github.com/scikit-hep/resample](https://github.com/scikit-hep/resample/blob/develop/src/rcont.c#L47). I am currently to map that to Python and will compare with @mdhaber implementation.

> In any case, he references an [earlier paper](https://www.jstor.org/stable/2347219) that does give the PMF explicitly.

Ah, cool.

**Comment by HDembinski** [comment]
> In a follow-up PR, I suggested that we could speed up the algorithm using Pythran, or we could replace it with the Patefield algorithm (or choose based on properties of the table), then speed it up with Pythran. We could go with C/C++ instead of Pythran, but in that case I can't help as much.

I did some benchmarks already and both the shuffling and Patefield's algorithm are useful to have. They have different time complexity, not to mention the space complexity of the shuffling algorithm which makes this algorithm unfeasible for tables with total number of entries N in the millions. Basically, the shuffling algorithm has O(N) complexity, but is "constant"  with respect to the number of cells K (eventually you see cache misses for large K), while Patefield scales with K x log(N). Patefield wins when N is large, but typically loses if N is small (in the hundreds or less).

I think we should support both with a keyword "method" or similar. If method is None, we can select the more appropriate algorithm empirically based on the inputs.

**Comment by HDembinski** [comment]
I am working on a prototype as discussed. I have never tried to compile scipy from sources and I need some help.

I am running MacOS 12.6 on an Apple M1 Pro.

I followed instructions here: https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#development-workflow

I can build and run the tests with `python dev.py test -v`, but I am seeing tests failing that are unrelated to my changes.

```
FAILED scipy/fft/_pocketfft/tests/test_real_transforms.py::test_dst4_definition_ortho[1-float32] - AssertionError:
FAILED scipy/linalg/tests/test_solvers.py::test_solve_discrete_are - AssertionError:
FAILED scipy/ndimage/tests/test_datatypes.py::test_uint64_max - AssertionError
FAILED scipy/sparse/linalg/_isolve/tests/test_iterative.py::test_maxiter_worsening[gmres] - AssertionError
FAILED scipy/special/tests/test_hyp2f1.py::TestHyp2f1::test_a_b_neg_int_after_euler_hypergeometric_transformation[hyp2f1_test_case0] - AssertionError:
FAILED scipy/special/tests/test_hyp2f1.py::TestHyp2f1::test_region4[hyp2f1_test_case30] - AssertionError:
FAILED scipy/special/tests/test_hyp2f1.py::TestHyp2f1::test_region4[hyp2f1_test_case34] - AssertionError:
FAILED scipy/special/tests/test_hyp2f1.py::TestHyp2f1::test_region4[hyp2f1_test_case47] - AssertionError:
FAILED scipy/special/tests/test_hyp2f1.py::TestHyp2f1::test_region5[hyp2f1_test_case3] - AssertionError:
FAILED scipy/stats/tests/test_boost_ufuncs.py::test_stats_boost_ufunc[_beta_ppf-args1-0.5] - RuntimeWarning: overflow encountered in _beta_ppf
FAILED scipy/stats/tests/test_continuous_basic.py::test_cont_basic[500-200-ncf-arg74] - RuntimeWarning: overflow encountered in _ncf_ppf
FAILED scipy/stats/tests/test_continuous_basic.py::test_cont_basic[500-200-nct-arg75] - RuntimeWarning: overflow encountered in _nct_ppf
FAILED scipy/stats/tests/test_continuous_basic.py::test_cont_basic[500-200-ncx2-arg76] - RuntimeWarning: overflow encountered in _ncx2_cdf
FAILED scipy/stats/tests/test_continuous_basic.py::test_methods_with_lists[ncf-args75-ppf] - RuntimeWarning: overflow encountered in _ncf_ppf
FAILED scipy/stats/tests/test_continuous_basic.py::test_methods_with_lists[ncf-args75-isf] - RuntimeWarning: overflow encountered in _ncf_isf
FAILED scipy/stats/tests/test_discrete_basic.py::test_discrete_basic[skellam-arg17-True] - RuntimeWarning: overflow encountered in _ncx2_cdf
FAILED scipy/stats/tests/test_discrete_basic.py::test_moments[skellam-arg17] - RuntimeWarning: overflow encountered in _ncx2_cdf
FAILED scipy/stats/tests/test_discrete_distns.py::test_skellam_gh11474 - RuntimeWarning: overflow encountered in _ncx2_cdf
FAILED scipy/stats/tests/test_discrete_distns.py::TestNCH::test_nchypergeom_wallenius_naive - RuntimeWarning: divide by zero encountered in mean (vector...
FAILED scipy/stats/tests/test_distributions.py::TestSkellam::test_cdf - RuntimeWarning: overflow encountered in _ncx2_cdf
FAILED scipy/stats/tests/test_distributions.py::TestExpect::test_skellam - RuntimeWarning: overflow encountered in _ncx2_cdf
FAILED scipy/stats/tests/test_distributions.py::test_ncx2_tails_ticket_955 - RuntimeWarning: overflow encountered in _ncx2_cdf
FAILED scipy/stats/tests/test_distributions.py::test_ncx2_gh12731 - RuntimeWarning: overflow encountered in _ncx2_cdf
FAILED scipy/stats/tests/test_distributions.py::test_ncx2_gh8665 - RuntimeWarning: overflow encountered in _ncx2_sf
========================== 24 failed, 37632 passed, 2207 skipped, 12238 deselected, 135 xfailed, 13 xpassed in 332.95s (0:05:32) ==========================
```
I tried `conda develop .` and `import scipy` from another directory, but got this error
```
Python 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:38:29) [Clang 13.0.1 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import scipy
Traceback (most recent call last):
  File "/Users/hdembinski/Extern/scipy/scipy/__init__.py", line 154, in <module>
    from scipy._lib._ccallback import LowLevelCallable
  File "/Users/hdembinski/Extern/scipy/scipy/_lib/_ccallback.py", line 1, in <module>
    from . import _ccallback_c
ImportError: cannot import name '_ccallback_c' from 'scipy._lib' (/Users/hdembinski/Extern/scipy/scipy/_lib/__init__.py)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/hdembinski/Extern/scipy/scipy/__init__.py", line 159, in <module>
    raise ImportError(msg) from e
ImportError: The `scipy` install you are using seems to be broken, (extension modules cannot be imported), please try reinstalling.
```
I tried reinstalling, but no change.

**Comment by tupui** [comment]
I would not really bother about failing tests if they are not related to your changes. The majority seem to be passing which should be fine for what you want to do.

`python dev.py build` is installing SciPy in the build folder. You then need to export this path in your PYTHONPATH. e.g. `export PYTHONPATH=/Users/.../scipy-repo/build-install/lib/python3.10/site-packages:`. This is what is happening when you do `python dev.py shell` for instance. Don't play with `conda develop .` this is deprecated and also change a pth file.

**Comment by WarrenWeckesser** [comment]
"Shuffling algorithm" is a much better name!

My understanding is that Patefield's algorithm is limited to 2-d tables.  An advantage of the shuffling algorithm is that it is easy to extend it to n-d tables.  That's what I implemented in `numerand.random_table` and in my C version.

A `method` parameter (with automatic selection of the "best" method when `method` is None) makes sense, as we already have three methods:

* Shuffling algorithm (can be used for n-d tables)
* Patefield (limited to 2-d tables)
* Hypergeometric distribution (limited to 2x2 tables).

The signature that I used for `numerand.random_table` is

    def random_table(*sums, size=None, method=None, rng=None)

For an n-d table, `len(sums)` is `n`; `n` is the number of categorical variables.  `sums[k]` is a 1-d sequence, with `len(sums[k])` being the number of levels of the k-th variable, and `sums[k][j]` being the frequency (i.e. number of occurrences) of the j-th level of the k-th variable.   For a 2-d table, this means that `sums[0]` is the sum along axis 1, and `sums[1]` is the sum along axis 0.  That might seem backwards, but it agrees with the n-d case, where `sums[k]` is the sum along all axes *except* `k`.  To avoid confusion, I think it is helpful to first describe `sums[k]` as the frequencies of the levels of the `k`-th categorical variable before mentioning "marginal sums".

E.g.

```
In [75]: numerand.random_table([10, 20, 30], [25, 35])
Out[75]: 
array([[ 4,  6],
       [10, 10],
       [11, 19]])

In [76]: numerand.random_table([10, 20, 30], [25, 35], size=3)
Out[76]: 
array([[[ 3,  7],
        [10, 10],
        [12, 18]],

       [[ 4,  6],
        [ 6, 14],
        [15, 15]],

       [[ 3,  7],
        [ 6, 14],
        [16, 14]]])

In [77]: numerand.random_table([5, 11], [12, 4], [8, 8])  # Table with shape (2, 2, 2)
Out[77]: 
array([[[3, 1],
        [1, 0]],

       [[3, 5],
        [1, 2]]])

In [78]: numerand.random_table([5, 11], [12, 4], [8, 8], size=3)
Out[78]: 
array([[[[1, 2],
         [1, 1]],

        [[5, 4],
         [1, 1]]],


       [[[2, 1],
         [1, 1]],

        [[4, 5],
         [1, 1]]],


       [[[3, 1],
         [0, 1]],

        [[4, 4],
         [1, 2]]]])
```

**Comment by HDembinski** [comment]
> Don't play with conda develop . this is deprecated and also change a pth file.

Ok, thank you for the pointer, I will try your suggestion, but `conda develop .` is the recommended way in https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#development-workflow

**Comment by tupui** [comment]
> Ok, thank you for the pointer, I will try your suggestion, but `conda develop .` is the recommended way in [scipy.github.io/devdocs/dev/contributor/conda_guide.html#conda-guide](https://scipy.github.io/devdocs/dev/contributor/conda_guide.html#conda-guide)

Yes this was not updated. The path is not correct as the build is not in `.` but in the build folder. We might have soon in-place builds with Meson and that will fix that.

**Comment by HDembinski** [comment]
Sorry, no progress here, because I cannot test the new code I wrote due to the errors that I mentioned. I cannot recover from the error that I posted above. I am a bit disappointed by how difficult it is to set up scipy for development. Is this a bad time due to the switch to meson or is this normal?

For example, why is the folder `scipy` not a put in a subfolder called `src`, as is standard in many projects?

**Comment by HDembinski** [comment]
I deleted my conda environment and recreated it from scratch. I deleted all build produts with `git clean -x -d -f`. When trying to build, I now get this error.
```
scipy/meson.build:40:0: ERROR: Command "/Users/hdembinski/mambaforge/envs/scipy-dev/bin/python3.10 -c import os; os.chdir(".."); import numpy; print(numpy.get_include())" failed with status 1.
```

**Comment by HDembinski** [comment]
Here is the meson log, it claims that numy is not found, but if I call the offending command
`/Users/hdembinski/mambaforge/envs/scipy-dev/bin/python3.10 -c 'import os; os.chdir(".."); import numpy; print(numpy.get_include())'` myself, it succeeds.
[meson-log.txt](https://github.com/scipy/scipy/files/9792348/meson-log.txt)

**Comment by HDembinski** [comment]
@mdhaber I submitted the PR. Docs are there, basic unit tests are there. No benchmarks yet, because we decided to not go for speed in the first iteration. No implementation of patefield's algorithm either, I just implemented the shuffling algorithm. I should now send an email to the scipy mailing list, right?

**Comment by mdhaber** [comment]
Yes, sounds good. Glad you got your build working!

**Comment by HDembinski** [comment]
Reading the Boyett paper, I realize that the "shuffling" algorithm is "Boyett's algorithm". I also implemented the pmf now that @rkern discovered in Boyett's paper.

**Comment by HDembinski** [comment]
@mdhaber Once we are done with this, I would like to contribute the USP test to scipy.stats.contingency as well, which is supposed to perform better than scipy.stats.contingency.chi2_contingency.

**Comment by mdhaber** [comment]
That sounds great. The paper describes the test as a permutation test, but that is just the trivial way of generating the contingency tables under the null hypothesis, which is what `random_table` will do. Could you submit the PR for the Patefield algorithm first? It would be nice to sample all the contingency tables in one call to `rvs`, but that can be very memory-intensive with `boyett` (I crashed my computer once...).
